src/checkpoint.hh

```diff
@@ -0,0 +1,1335 @@
+/* {{{
+
+    This file is part of gapc (GAPC - Grammars, Algebras, Products - Compiler;
+      a system to compile algebraic dynamic programming programs)
+
+    Copyright (C) 2008-2011  Georg Sauthoff
+         email: gsauthof@techfak.uni-bielefeld.de or gsauthof@sdf.lonestar.org
+
+    This program is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+    * Author: fymue
+}}} */
+
+#ifndef SRC_CHECKPOINT_HH_
+#define SRC_CHECKPOINT_HH_
+
+#include <string>
+#include <vector>
+#include <list>
+#include <utility>
+#include <sstream>
+#include "printer.hh"
+#include "symbol.hh"
+#include "statement/table_decl.hh"
+#include "type/base.hh"
+
+// default checkpointing interval
+// (modify here if you want to change it)
+#define DEFAULT_CP_INTERVAL_SEC     3600
+#define DEFAULT_CP_INTERVAL_MIN     (DEFAULT_CP_INTERVAL_SEC / 60)
+
+typedef hashtable<std::string, Symbol::NT*> nt_tables;
+
+namespace Printer {
+/*
+  contains methods that handle the insertion of the checkpointing
+  routine into the generated header file;
+  this class also extends Base to get access to indent-related methods
+*/
+class Checkpoint : public Base {
+ private:
+  /*
+     true if table types are wrapped in List_Ref;
+     this information is required whenever the String type
+     is part of a table's datatype, because tables containting
+     Strings need to be additionally processed after deserialization,
+     which can only work if the Strings are wrapped in a List_Ref object
+  */
+  bool list_ref;
+
+  std::vector<std::string> string_type_accessors, block_type_accessors,
+                           subseq_type_accessors;
+
+  // currently supported/serializable GAPC-internal datatypes (2023-02-20)
+  const std::vector<Type::Type>
+  SUPPORTED_TYPES = {Type::Type::VOID, Type::Type::INTEGER,
```

Comment:
Ah, that's one of those cases where the table type for some reason isn't enumerated properly. In this case, the table type was enumerated as `Type::Usage`, whose `base` member then pointed to the actual type `char`. I don't know when/why that happens, but that case is handled now.

Also since `char` is primitive, serializing `char` tables is not going to require any further modifications to the the checkpointing mechanism.